assets/javascripts/discourse/services/chat-message-visibility-observer.js

```diff
@@ -0,0 +1,35 @@
+import Service from "@ember/service";
+
+export default class ChatMessageVisibilityObserver extends Service {
+  constructor() {
+    super(...arguments);
+    this._setup();
+  }
+
+  _setup() {
+    const options = {
+      root: document,
+      rootMargin: "-10px",
+    };
+
+    const callback = (entries) => {
+      entries.forEach((entry) => {
+        entry.target.dataset.visible = entry.isIntersecting;
+      });
+    };
+
+    this.observer = new IntersectionObserver(callback, options);
```

Comment:
I wonder if it would be cleaner to make `callback` a method on the service, rather than an anonymous function?